File: php/testcode/xxe_042.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_xxe_xml_signature_verify
function xxe042(BenchmarkRequest $req): BenchmarkResponse {
    $xml = $req->bodyStr();
    $sig = $req->header('X-XML-Signature');
    if (!openssl_verify($xml, base64_decode($sig), file_get_contents('/etc/app/xml.pub'))) { // vuln-code-snippet safe-line php_xxe_xml_signature_verify
        return BenchmarkResponse::badRequest('invalid sig');
    }
    libxml_disable_entity_loader(true);
    $data = simplexml_load_string($xml);
    return BenchmarkResponse::ok((string)$data);
}
// vuln-code-snippet end php_xxe_xml_signature_verify

File: ruby/testcode/xxe_042.rb
require_relative 'shared'
require 'nokogiri'

# vuln-code-snippet start ruby_xxe_schema_validate_first
def parse_schema_validate_first_handler(req)
  xml = req.post('xml')
  xsd = Nokogiri::XML::Schema.new(File.read('schema.xsd'))
  doc = Nokogiri::XML.parse(xml) { |c| c.nonet.nodtdload } # vuln-code-snippet safe-line ruby_xxe_schema_validate_first
  errors = xsd.validate(doc)
  raise 'Invalid XML' unless errors.empty?
  BenchmarkResponse.json({ root: doc.root&.name })
end
# vuln-code-snippet end ruby_xxe_schema_validate_first
